mbf_abstract_nav/src/move_base_action.cpp

```diff
@@ -544,36 +509,57 @@ void MoveBaseAction::actionRecoveryDone(
-  if (!replanning_ || action_state_ != EXE_PATH)
-    return; // replan only while following a path and if replanning is enabled (can be disabled by dynamic reconfigure)
+  ros::Duration update_period(0.001);
+  ros::Time last_replan_time(0.0);
 
-  if (state == actionlib::SimpleClientGoalState::SUCCEEDED)
+  while (ros::ok())
   {
-    ROS_DEBUG_STREAM_NAMED("move_base", "Replanning succeeded; sending a goal to \"exe_path\" with the new plan");
-    exe_path_goal_.path = result->path;
-    mbf_msgs::ExePathGoal goal(exe_path_goal_);
-    action_client_exe_path_.sendGoal(
-        goal,
-        boost::bind(&MoveBaseAction::actionExePathDone, this, _1, _2),
-        boost::bind(&MoveBaseAction::actionExePathActive, this),
-        boost::bind(&MoveBaseAction::actionExePathFeedback, this, _1));
-  }
+    // replan only while following a path and if replanning is enabled (can be disabled by dynamic reconfigure)
+    bool keep_replanning = !replanning_period_.isZero() && action_state_ == EXE_PATH && dist_to_goal_ > 0.1;
 
-  replanning_mtx_.lock();
-  replanning_rate_.sleep();
-  replanning_mtx_.unlock();
+    if (!action_client_get_path_.getState().isDone())
+    {
+      if (action_client_get_path_.waitForResult(update_period))
+      {
+        actionlib::SimpleClientGoalState state = action_client_get_path_.getState();
+        mbf_msgs::GetPathResultConstPtr result = action_client_get_path_.getResult();
+        if (state == actionlib::SimpleClientGoalState::SUCCEEDED && keep_replanning)
+        {
+          ROS_DEBUG_STREAM_NAMED("move_base", "Replanning succeeded; sending a goal to \"exe_path\" with the new plan");
+          exe_path_goal_.path = result->path;
+          mbf_msgs::ExePathGoal goal(exe_path_goal_);
+          action_client_exe_path_.sendGoal(goal, boost::bind(&MoveBaseAction::actionExePathDone, this, _1, _2),
+                                           boost::bind(&MoveBaseAction::actionExePathActive, this),
+                                           boost::bind(&MoveBaseAction::actionExePathFeedback, this, _1));
+        }
+        else
+        {
+          ROS_DEBUG_STREAM_NAMED("move_base",
+                                 "Replanning failed with error code " << result->outcome << ": " << result->message);
+        }
+      }
+      else
+      {
+        // keep waiting for planning to complete (we already waited update_period in waitForResult)
+        continue;
+      }
+    }
 
-  if (!replanning_ || action_state_ != EXE_PATH)
-    return; // another chance to stop replannings after waiting for replanning period
+    if (!keep_replanning)
+    {
+      update_period.sleep();
+      continue;
+    }
 
-  ROS_DEBUG_STREAM_NAMED("move_base", "Next replanning cycle, using the \"get_path\" action!");
-  action_client_get_path_.sendGoal(
-      get_path_goal_,
-      boost::bind(&MoveBaseAction::actionGetPathReplanningDone, this, _1, _2)); // replanning
+    if (ros::Time::now() - last_replan_time >= replanning_period_)
```

Comment:
does it make sense to check the state of action_client_get_path_ - I'm thinking that if the user picks a replanning period shorted than the planners response time we will always preempt the planner

Reply:
just personal preference but it seems better readable if we make the ifs else ifs:

```
if(!action_client_get_path_.getState().isDone()){...}
else if(!keep_replanning) {...}
else if(ros::Time::now() - last_replan_time >= replanning_period_) {...}
```
in this manner we would also resolve the issue with preempting the planer

Reply:
> does it make sense to check the state of action_client_get_path_ - I'm thinking that if the user picks a replanning period shorted than the planners response time we will always preempt the planner

nope,,,  we will continue [here](https://github.com/magazino/move_base_flex/pull/267/files#diff-6b8a12259c3d11b5f68c080c8a6d7a690aecea38015c66366f1d559b16bd4acbR546) and so not send a new goal

Reply:
ah yea you're right - my bad